src/pages/api/revalidate.ts

```diff
@@ -0,0 +1,79 @@
+import { NextApiRequest, NextApiResponse } from 'next/types'
+import { StatusCodes } from 'http-status-codes'
+import { validateToken } from 'lib/api-validate-token'
+import { cachedGetProductData } from 'lib/get-product-data'
+
+/**
+ * Accepts a POST request with a product slug, triggers revalidation for all of a product's paths
+ * specified in its latest nav data.
+ */
+export default validateToken(
+	async function handler(request: NextApiRequest, response: NextApiResponse) {
+		if (request.method !== 'POST') {
+			response.status(StatusCodes.NOT_FOUND)
+			return
+		}
+
+		console.log(request.body)
+		const { product } = request.body
+
+		if (!product) {
+			response
+				.status(StatusCodes.BAD_REQUEST)
+				.json({ error: 'No product provided.' })
+
+			return
+		}
+
+		const productData = cachedGetProductData(product)
+
+		const navDataPrefixes = productData.rootDocsPaths.map(
+			({ navDataPrefix, path }) => {
+				return { navDataPrefix, path }
+			}
+		)
+
+		// fetch the latest nav data, which will be used to construct paths to revalidate
+		const navDataFiles = await Promise.all(
+			navDataPrefixes.map(({ navDataPrefix, path }) => {
+				const prefix = navDataPrefix ?? path
+
+				return fetch(
+					`https://content.hashicorp.com/api/content/${product}/nav-data/latest/${prefix}`
+				)
+					.then((res) => res.json())
+					.then(({ result }) => [path, result.navData])
+			})
+		)
+
+		const revalidatePromises = []
+
+		// iterates through nav data recursively and creates a revalidate request
+		function revalidateNavNodes(basePath, nodes) {
+			nodes.forEach(({ path, routes }) => {
+				if (routes) {
+					revalidateNavNodes(basePath, routes)
+					return
+				}
+
+				if (path) {
+					const pathToRevalidate = `/${product}/${basePath}/${path}`
+
+					console.log('[revalidate]', pathToRevalidate)
+					revalidatePromises.push(response.revalidate(pathToRevalidate))
+				}
+			})
+		}
+
+		// iterate over each nav node and revalidate its path
+		navDataFiles.forEach(([basePath, navData]) => {
+			revalidateNavNodes(basePath, navData)
+		})
+
+		// wait for everything to get revalidated
+		await Promise.allSettled(revalidatePromises)
```

Comment:
🙋does this need to be wrapped in a `try {} catch {}` so that we always return a response even if this Promise gets rejected? In local development I'm not getting a response from the server for this endpoint.

Reply:
That's a great question. I think with `Promise.allSettled` we would have to iterate through the results here and check for a `rejected` status, as it doesn't reject itself. So the `await` on its own _should_ be okay here.

I'm going to add a TODO to add retries and/or additional resiliency here though. Given the volume of requests being sent out from this, it seems likely that we'll see a network hiccup or two.